File: android/src/main/java/com/example/ivs_broadcaster/streams/StageListener.java
package com.example.ivs_broadcaster.streams;

import android.annotation.SuppressLint;
import android.graphics.Color;
import android.os.Build;
import android.util.Log;
import android.view.SurfaceView;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;

import com.amazonaws.ivs.broadcast.BroadcastException;
import com.amazonaws.ivs.broadcast.ImagePreviewSurfaceView;
import com.amazonaws.ivs.broadcast.ImagePreviewView;
import com.amazonaws.ivs.broadcast.ParticipantInfo;
import com.amazonaws.ivs.broadcast.Stage;
import com.amazonaws.ivs.broadcast.StageRenderer;
import com.amazonaws.ivs.broadcast.StageStream;
import com.amazonaws.ivs.chat.messaging.requests.SendMessageRequest;
import com.example.ivs_broadcaster.chats.StageChat;
import com.example.ivs_broadcaster.views.RemoteViewFactory;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.UUID;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.EventChannel;

@SuppressLint("NewApi")
public class StageListener implements StageRenderer {
    StageChat stageChat;
    FlutterPlugin.FlutterPluginBinding binding;
    HashMap map = new HashMap();
    EventChannel.EventSink sink;
    ArrayList<String> joinedParticipants = new ArrayList<>();
    ArrayList<String> viewsParticipants = new ArrayList<>();
    ArrayList<String> leftParticipants = new ArrayList<>();

    public StageListener(FlutterPlugin.FlutterPluginBinding binding, EventChannel.EventSink sink, StageChat stageChat) {
        this.sink = sink;
        this.binding = binding;
        this.stageChat = stageChat;
        Log.d("StageListener", "Init " + sink + " ");
    }

    @Override
    public void onError(@NonNull BroadcastException exception) {
        StageRenderer.super.onError(exception);
        Log.d("StageListener", exception.toString());
    }

    @Override
    public void onConnectionStateChanged(@NonNull Stage stage, @NonNull Stage.ConnectionState state, @Nullable BroadcastException exception) {
        StageRenderer.super.onConnectionStateChanged(stage, state, exception);
        map.put("connection_state_changed", state.name());
        if (sink != null) sink.success(map);
    }

    @Override
    public void onParticipantJoined(@NonNull Stage stage, @NonNull ParticipantInfo participantInfo) {
        StageRenderer.super.onParticipantJoined(stage, participantInfo);
        String message = String.format("<>^S^E^R^V^E^R<>::dev::{\"type\":\"participantJoined\",\"category\":\"liveRoom\",\"data\":{\"participantId\":\"%s\"}}", participantInfo.participantId);
        try {
            if (stageChat.room != null) {
                stageChat.room.sendMessage(new SendMessageRequest(message));
            }
            Log.d("StageChat", "Message sent");
        } catch (Exception e) {
            Log.d("StageChat", e.getMessage());
        }
        joinedParticipants.removeIf(element -> element.equals(participantInfo.participantId));
        joinedParticipants.add(participantInfo.participantId);
        map.put("joined_participants", joinedParticipants);
        if (sink != null) sink.success(map);
    }

    @Override
    public void onParticipantLeft(@NonNull Stage stage, @NonNull ParticipantInfo participantInfo) {
        StageRenderer.super.onParticipantLeft(stage, participantInfo);
        String message = String.format("<>^S^E^R^V^E^R<>::dev::{\"type\":\"participantLeft\",\"category\":\"liveRoom\",\"data\":{\"participantId\":\"%s\"}}", participantInfo.participantId);
        try {
            if (stageChat.room != null) {
                stageChat.room.sendMessage(new SendMessageRequest(message));
            }
            Log.d("StageChat", "Message sent left");
        } catch (Exception e) {
            Log.d("StageChat", e.getMessage());
        }
        leftParticipants.removeIf(element -> element.equals(participantInfo.participantId));
        leftParticipants.add(participantInfo.participantId);
        map.put("left_participants", leftParticipants);
        if (sink != null) sink.success(map);
    }

    @Override
    public void onParticipantPublishStateChanged(@NonNull Stage stage, @NonNull ParticipantInfo participantInfo, @NonNull Stage.PublishState publishState) {
        StageRenderer.super.onParticipantPublishStateChanged(stage, participantInfo, publishState);
        map.put("participant_publish_state_changed", publishState.name());
        if (sink != null) sink.success(map);
    }

    @Override
    public void onParticipantSubscribeStateChanged(@NonNull Stage stage, @NonNull ParticipantInfo publishingParticipantInfo, @NonNull Stage.SubscribeState subscribeState) {
        StageRenderer.super.onParticipantSubscribeStateChanged(stage, publishingParticipantInfo, subscribeState);
        map.put("participant_subscribe_state_changed", subscribeState.name());
        if (sink != null) sink.success(map);
    }

    @Override
    public void onStreamsAdded(@NonNull Stage stage, @NonNull ParticipantInfo participantInfo, @NonNull List<StageStream> streams) {
        StageRenderer.super.onStreamsAdded(stage, participantInfo, streams);

        for (int i = 0; i < streams.size(); i++) {
            if (streams.get(i).getStreamType() == StageStream.Type.VIDEO) {
                String viewId = UUID.randomUUID().toString() + "_" + streams.get(i).getDevice().getDescriptor().deviceId;
//                View view = streams.get(i).getPreviewSurfaceView();
                View view = streams.get(i).getPreviewSurfaceView();
                final boolean result = binding.getPlatformViewRegistry().registerViewFactory(viewId, new RemoteViewFactory(view));
                Log.d("PlatformView", String.valueOf(result));
                Log.d("PlatformView", String.valueOf(streams.get(i).getDevice().isValid()));
                viewsParticipants.removeIf(element -> element.equals(viewId));
                viewsParticipants.add(viewId);
            }
        }
        map.put("stream_added", viewsParticipants);
        if (sink != null) sink.success(map);
    }

    @Override
    public void onStreamsRemoved(@NonNull Stage stage, @NonNull ParticipantInfo participantInfo, @NonNull List<StageStream> streams) {
        StageRenderer.super.onStreamsRemoved(stage, participantInfo, streams);
        for (int i = 0; i < streams.size(); i++) {
            if (streams.get(i).getStreamType() == StageStream.Type.VIDEO) {
                String deviceId = streams.get(i).getDevice().getDescriptor().deviceId;
                viewsParticipants.removeIf(element -> element.contains(deviceId));
            }
        }
        map.put("stream_removed", viewsParticipants);
        if (sink != null) sink.success(map);
    }

    @Override
    public void onStreamsMutedChanged(@NonNull Stage stage, @NonNull ParticipantInfo participantInfo, @NonNull List<StageStream> streams) {
        StageRenderer.super.onStreamsMutedChanged(stage, participantInfo, streams);
        ArrayList<String> streamsMutedChanged = new ArrayList<String>();
        for (int i = 0; i < streams.size(); i++) {
            streamsMutedChanged.add(streams.get(i).getDevice().getDescriptor().urn);
        }
        map.put("streams_muted_changed", streamsMutedChanged);
        if (sink != null) sink.success(map);
    }
}

File: ios/Classes/streams/StageListener.swift
import Flutter
import AmazonIVSBroadcast
import AmazonIVSChatMessaging

public class StageListener: NSObject, IVSStageRenderer {
    private var map: [String: Any] = [:]
    public var stageSink: FlutterEventSink?
    var joinedParticipants: [String] = []
    var leftParticipants: [String] = []
    var viewsParticipants: [String] = []
    private var stageChat: StageChat
    private var binding: FlutterPluginRegistrar

    init(stageSink: FlutterEventSink? = nil, stageChat: StageChat, binding: FlutterPluginRegistrar) {
        self.binding = binding
        self.stageSink = stageSink
        self.stageChat = stageChat
        print("StageListener init");
    }
    
    public func stage(_ stage: IVSStage, participantDidJoin participant: IVSParticipantInfo) {
        let participantId = participant.participantId
        let message = "<>^S^E^R^V^E^R<>::dev::{\"type\":\"participantJoined\",\"category\":\"liveRoom\",\"data\":{\"participantId\":\"\(participantId)\"}}"
        stageChat.chatRoom?.sendMessage(with: SendMessageRequest(content: message))
        print("StageListener message sent \(message)")
        joinedParticipants.removeAll { $0 == participant.participantId }
        joinedParticipants.append(participant.participantId)
        map["joined_participants"] = joinedParticipants
        stageSink?(map)
        print("StageListener joined")
    }
    
    public func stage(_ stage: IVSStage, participantDidLeave participant: IVSParticipantInfo) {
        let participantId = participant.participantId
        let message = "<>^S^E^R^V^E^R<>::dev::{\"type\":\"participantLeft\",\"category\":\"liveRoom\",\"data\":{\"participantId\":\"\(participantId)\"}}"
        stageChat.chatRoom?.sendMessage(with: SendMessageRequest(content: message))
        print("StageListener message sent \(message)")
        leftParticipants.removeAll { $0 == participant.participantId }
        leftParticipants.append(participant.participantId)
        map["left_participants"] = leftParticipants
        stageSink?(map)
        print("StageListener leaved")
    }
    
 
    public func stage(_ stage: IVSStage, participant: IVSParticipantInfo, didChange publishState: IVSParticipantPublishState) {
        map["participant_publish_state_changed"] =  publishState.rawValue
        stageSink?(map)
        print("StageListener publish state")
    }
    
    public func stage(_ stage: IVSStage, participant: IVSParticipantInfo, didChange subscribeState: IVSParticipantSubscribeState) {
        map["participant_subscribe_state_changed"] =  subscribeState.rawValue
        stageSink?(map)
        print("StageListener subscribe state")
    }
    
    public func stage(_ stage: IVSStage, participant: IVSParticipantInfo, didAdd streams: [IVSStageStream]) {
        for stream in streams {
            if let imageDevice = stream.device as? IVSImageDevice {
                let viewId = UUID().uuidString + "_" + stream.device.descriptor().urn;
                do {
                    binding.register(RemoveViewFactory(view: try imageDevice.previewView()), withId: viewId)
                    print("StageListener view is registered")

                }catch {
                    print("StageListener view not register")
                }
                viewsParticipants.removeAll { $0 == viewId }
                viewsParticipants.append(viewId)
            }
        }
        map["stream_added"] = viewsParticipants;
        stageSink?(map)
        print("StageListener stream added")

    }
    
    public func stage(_ stage: IVSStage, participant: IVSParticipantInfo, didRemove streams: [IVSStageStream]) {
        for stream in streams {
            if let imageDevice = stream.device as? IVSImageDevice {
                let viewId = stream.device.descriptor().urn;
                viewsParticipants.removeAll { $0.contains(viewId)}
            }
        }
        map["stream_removed"] = viewsParticipants;
        stageSink?(map)
        print("StageListener stream removed")
    }
    
    public func stage(_ stage: IVSStage, didChange connectionState: IVSStageConnectionState, withError error: (any Error)?) {
        print("StageListener: \(error) \(connectionState) \(stage)")
    }
    
    public func stage(_ stage: IVSStage, participant: IVSParticipantInfo, didChangeMutedStreams streams: [IVSStageStream]) {
        var streamsMutedChanged: [String] = []
        for stream in streams {
            let urn = stream.device.descriptor().urn
            streamsMutedChanged.append(urn)
        }
        map["streams_muted_changed"] = streamsMutedChanged
        stageSink?(map)
        print("StageListener stream muted")

    }
}
